### PR TITLE
Standalone PR links first epic

### LINKED ISSUE #80: Standalone PR links first epic
When standalone PR is created and links multiple issues, iterate issues and:
* Use 1st epic and use it as a parent
* Create issue without a parent otherwise

Depends on #77 and #78

### PR DESCRIPTION
**Pull request by sebastien-marichal**
Fixes #80

### CODE REVIEW

File: PullRequestCreated/PullRequestCreated.ts [hunk 1]
```diff
@@ -45,7 +55,7 @@ class PullRequestCreated extends OctokitAction {
   }
 
   private findMentionedIssues(pr: PullRequest): string[] {
-    return pr.body?.match(JIRA_ISSUE_PATTERN) || [];
+    return [...new Set(pr.body?.match(JIRA_ISSUE_PATTERN) || [])];
```
Comment (sebastien-marichal) [inline PullRequestCreated/PullRequestCreated.ts:58]: `.Distinct()`

File: PullRequestCreated/PullRequestCreated.ts [hunk 2]
```diff
@@ -30,7 +30,17 @@ class PullRequestCreated extends OctokitAction {
   private async newIssueParameters(pr: PullRequest): Promise<IssueParameters> {
     const mentionedIssues = this.findMentionedIssues(pr);
     console.log(`Found mentioned issues: ${mentionedIssues}`);
-    const parent = mentionedIssues.length === 1 ? await this.jira.getIssue(mentionedIssues[0]) : null;
+    let parent = mentionedIssues.length === 1 ? await this.jira.getIssue(mentionedIssues[0]) : null;
+    if (mentionedIssues.length > 1) {
+      console.log('Looking for an Epic.');
+      for (const issueKey of mentionedIssues) {
+        const issue = await this.jira.getIssue(issueKey);
+        if (issue != null && issue.fields.issuetype.name !== 'Sub-Task') {
```
Comment (sebastien-marichal) [inline PullRequestCreated/PullRequestCreated.ts:38]: As discussed [here](https://github.com/SonarSource/gh-action-lt-backlog/pull/104#discussion_r1697994366), I am taking the first non-`Sub-task`

File: PullRequestCreated/PullRequestCreated.ts [hunk 3]
```diff
@@ -30,7 +30,17 @@ class PullRequestCreated extends OctokitAction {
   private async newIssueParameters(pr: PullRequest): Promise<IssueParameters> {
     const mentionedIssues = this.findMentionedIssues(pr);
     console.log(`Found mentioned issues: ${mentionedIssues}`);
-    const parent = mentionedIssues.length === 1 ? await this.jira.getIssue(mentionedIssues[0]) : null;
+    let parent = mentionedIssues.length === 1 ? await this.jira.getIssue(mentionedIssues[0]) : null;
+    if (mentionedIssues.length > 1) {
+      console.log('Looking for an Epic.');
+      for (const issueKey of mentionedIssues) {
+        const issue = await this.jira.getIssue(issueKey);
+        if (issue != null && issue.fields.issuetype.name !== 'Sub-task') {
```
Comment (sebastien-marichal) [inline PullRequestCreated/PullRequestCreated.ts:38]: As discussed [here](https://github.com/SonarSource/gh-action-lt-backlog/pull/104#discussion_r1697994366), I am taking the first non-`Sub-task`

File: PullRequestCreated/PullRequestCreated.ts [hunk 4]
```diff
@@ -28,7 +28,17 @@ class PullRequestCreated extends OctokitAction {
   private async newIssueParameters(pr: PullRequest): Promise<IssueParameters> {
     const mentionedIssues = this.findMentionedIssues(pr);
     console.log(`Found mentioned issues: ${mentionedIssues}`);
-    const parent = mentionedIssues.length === 1 ? await this.jira.getIssue(mentionedIssues[0]) : null;
+    let parent = mentionedIssues.length === 1 ? await this.jira.getIssue(mentionedIssues[0]) : null;
+    if (mentionedIssues.length > 1) {
+      console.log('Looking for an Epic.');
+      for (const issueKey of mentionedIssues) {
+        const issue = await this.jira.getIssue(issueKey);
+        if (issue?.fields.issuetype.name !== 'Sub-task') {
```
Comment (sebastien-marichal) [inline PullRequestCreated/PullRequestCreated.ts:36]: As discussed [here](https://github.com/SonarSource/gh-action-lt-backlog/pull/104#discussion_r1697994366), I am taking the first non-`Sub-task`

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
We'll need to reorganize the code a little

File: PullRequestCreated/PullRequestCreated.ts (hunk 1 above)
Comment (sebastien-marichal) [reply PullRequestCreated/PullRequestCreated.ts:65]: It happened to me after adding the linked issue:
```
[PAVEL-1](https://sonarsource.atlassian.net/browse/PAVEL-1)
```

The issue will be found twice.

Comment (sebastien-marichal) [reply PullRequestCreated/PullRequestCreated.ts:65]: About not using Set directly, AFAIK, you cannot use it as an array.

File: PullRequestCreated/PullRequestCreated.ts [hunk 5]
```diff
@@ -28,7 +28,17 @@ class PullRequestCreated extends OctokitAction {
   private async newIssueParameters(pr: PullRequest): Promise<IssueParameters> {
     const mentionedIssues = this.findMentionedIssues(pr);
     console.log(`Found mentioned issues: ${mentionedIssues}`);
-    const parent = mentionedIssues.length === 1 ? await this.jira.getIssue(mentionedIssues[0]) : null;
+    let parent = mentionedIssues.length === 1 ? await this.jira.getIssue(mentionedIssues[0]) : null;
```
[the remark this answers is not in the document]
Comment (sebastien-marichal) [reply PullRequestCreated/PullRequestCreated.ts:40]: Left over from when I was trying to implement it with #104 and got confused...

File: PullRequestCreated/PullRequestCreated.ts (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply PullRequestCreated/PullRequestCreated.ts:70]: All we need is to iterate it

File: PullRequestCreated/PullRequestCreated.ts [hunk 6]
```diff
@@ -55,8 +55,19 @@ class PullRequestCreated extends OctokitAction {
     }
   }
 
+  private async firstNonSubTask(issues: string[]): Promise<any | null> {
+    console.log('Looking for a non-Sub-task ticket');
```
Comment (pavel-mikula-sonarsource) [inline PullRequestCreated/PullRequestCreated.ts:59]: Logging for this process is not here and in the caller site. I would like to have it on a consistent level instead.
It means either:
* Move `Found mentioned issues..` inside `findmentionedIssues`, and move `Parent issue....` here.
* Move `Looking for non-Sub-task ticket` to `newIssueParameters`

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM with some final polishing above and below

File: PullRequestCreated/PullRequestCreated.ts (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply PullRequestCreated/PullRequestCreated.ts:70]: The idea was to change return type and parameter of the `firstNonSubTask` to the Set itself. If possible.

**Pull request merged**